src/glyph.ts

```diff
@@ -5,20 +5,133 @@ import { Flow } from './tables';
 import { Element } from './element';
 import { BoundingBoxComputation } from './boundingboxcomputation';
 import { BoundingBox } from './boundingbox';
+import { Font } from './smufl';
+import { RenderContext } from './types/common';
+import { Stave } from './stave';
+
+export interface KeyProps {
+  stem_down_x_offset: number;
+  stem_up_x_offset: number;
+  key: string;
+  octave: number;
+  line: number;
+  int_value: number;
+  accidental: string;
+  code: string;
+  stroke: number;
+  shift_right: number;
+  displaced: boolean;
+}
+
+export interface TypeProps extends KeyProps {
+  getWidth(scale?: number): number;
+
+  code: string;
+  code_head: string;
+  stem: boolean;
+  rest: boolean;
+  flag: boolean;
+  stem_offset: number;
+  stem_up_extension: number;
+  stem_down_extension: number;
+  tabnote_stem_up_extension: number;
+  tabnote_stem_down_extension: number;
+  dot_shiftY: number;
+  line_above: number;
+  line_below: number;
+  beam_count: number;
+  code_flag_upstem: string;
+  code_flag_downstem: string;
+  position: string;
+}
+export interface DurationCode {
+  common: TypeProps;
+  type: Record<string, TypeProps>;
+}
 
-function processOutline(outline, originX, originY, scaleX, scaleY, outlineFns) {
-  let command;
-  let x;
-  let y;
+export interface GlyphProps {
+  code_head: string;
+  dot_shiftY: number;
+  position: string;
+  rest: boolean;
+  line_below: number;
+  line_above: number;
+  stem_up_extension: never;
+  stem_down_extension: never;
+  stem: never;
+  code: string;
+  code_flag_upstem: string;
+  code_flag_downstem: string;
+  flag: boolean;
+  width: number;
+  text: string;
+  tabnote_stem_down_extension: number;
+  tabnote_stem_up_extension: number;
+  beam_count: number;
+  duration_codes: Record<string, DurationCode>;
+  validTypes: Record<string, string>;
+  shift_y: number;
+
+  getWidth(a?: number): number;
+
+  getMetrics(): GlyphMetrics;
+}
+export interface GlyphOptions {
+  fontStack: Font[];
+  category?: string;
+}
+export interface GlyphMetrics {
+  width?: number;
+  height?: number;
+  x_min: number;
+  x_max: number;
+  x_shift: number;
+  y_shift: number;
+  scale: number;
+  ha?: number;
+  outline: string[];
+  font?: Font;
+}
+
+export interface FontGlyph {
+  x_min: number;
+  x_max: number;
+  y_min: number;
+  y_max: number;
+  ha: number;
+  o: string;
+  leftSideBearing?: number;
+  advanceWidth?: number;
+  cached_outline?: string[];
+}
+
+export interface GlyphLookup {
```

Comment:
This can be inline, no need for a separate type.

Reply:
ok